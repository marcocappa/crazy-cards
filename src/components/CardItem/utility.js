import React from 'react';
import { IconAnywhere, IconLiquid, IconStudent, IconCards } from '../Icons';

export const getCardIcon = (id) => {
  switch (id) {
    case 'card-0':
      return <IconStudent />;
    case 'card-1':
      return <IconAnywhere />;
    case 'card-2':
      return <IconLiquid />;
    default:
      return <IconCards />;
  }
};
