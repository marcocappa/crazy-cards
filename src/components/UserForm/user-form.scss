.crazy-card-user-form {
    max-width: 25rem;
    width: 100%;

    & label {
        display: flex;
        flex-direction: column;
        margin: 0 0 1rem;

        & span {
            display: block;
            margin-bottom: 0.25rem;
        }
    }

    &__actions {
        display: flex;
        justify-content: flex-end;
    }

    &__submit {
        cursor: pointer;
        background-color: #0f0a3a;
        border-radius: 0.5rem;
        border: 0.125rem solid transparent;
        color: #f4f4f4;
        display: inline-block;
        font-family: "Source Sans Pro", Arial, sans-serif;
        font-size: 0.875rem;
        padding: 0.625rem 1.125rem;
        text-decoration: none;
        transition: background-color 0.2s ease-in-out;

        &:hover {
            color: #ffffff;
            background-color: #403794;
        }
    }
}
