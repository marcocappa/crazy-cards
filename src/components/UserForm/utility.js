export const checkForm = ({
  firstname,
  lastname,
  title,
  dateOfBirth,
  employmentStatus,
  annualIncome,
  houseNumber,
  postcode,
}) =>
  !firstname ||
  !lastname ||
  !title ||
  !dateOfBirth ||
  !employmentStatus ||
  !annualIncome ||
  !houseNumber ||
  !postcode;
