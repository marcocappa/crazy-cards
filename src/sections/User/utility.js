import { ANNUAL_INCOME_LIMIT } from '../../common/constants';

/**
 * getFilteredCards function will filter an array of object based on user id
 * user employmentStatus and user annualIncome. It will return cards only applicable on user information
 *
 * @param {array} allCards
 * @param {object} user
 * @returns array of filtered cards
 */
export const getFilteredCards = (allCards, user) =>
  allCards.filter(
    (card) =>
      card.id === 'anywhere' ||
      (!('employmentStatus' in user) && card.id === 'student') ||
      (user.employmentStatus === 'Student' && card.id === 'student') ||
      (user.annualIncome > ANNUAL_INCOME_LIMIT && card.id === 'liquid')
  );

/**
 * getCheckboxList function will create a new array of object based on the cards
 *
 * @param {array} filteredCards
 * @returns array of new objects based on the filteredCards
 */
export const getCheckboxList = (filteredCards) =>
  filteredCards.map((card) => {
    return { id: card.id, value: card.id, label: card.name };
  });

/**
 * getCardListSelected function will return selected cards into a new array
 *
 * @param {array} filteredCards
 * @param {array} selectedIds | array with card ids selected
 * @returns array selectedCards
 */
export const getCardListSelected = (filteredCards, selectedIds) =>
  filteredCards.filter((card) => selectedIds.includes(card.id));

/**
 * getTotalCredit function will return totalCredit available based
 * on the cards selected
 *
 * @param {array} items | array with cards object containing creditAvailable
 * @returns array selectedCards
 */
export const getTotalCredit = (items) =>
  items.reduce((acc, item) => {
    return acc + item.creditAvailable;
  }, 0);
