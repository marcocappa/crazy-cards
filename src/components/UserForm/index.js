import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Select from 'react-select';
import DatePicker from 'react-datepicker';
import { checkForm, getToday } from './utility';
import './user-form.scss';

const titleOptions = [
  { value: 'Mr.', label: 'Mr.' },
  { value: 'Mrs.', label: 'Mrs.' },
  { value: 'Miss', label: 'Miss' },
];

const employmentStatusOptions = [
  { value: 'Student', label: 'Student' },
  { value: 'Part Time', label: 'Part Time' },
  { value: 'Full Time', label: 'Full Time' },
];

const initialUserState = {
  title: '',
  firstname: '',
  lastname: '',
  dateOfBirth: '',
  employmentStatus: '',
  annualIncome: 0,
  houseNumber: 0,
  postcode: '',
};

const UserForm = ({ onSubmit, today }) => {
  const [user, setUserData] = useState(initialUserState);
  const [error, setError] = useState(false);

  const handleChange = (field, value) => {
    setUserData({
      ...user,
      [field]: value,
    });
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    if (checkForm(user)) {
      setError(true);
      return;
    }
    setError(false);
    onSubmit(user);
  };

  return (
    <form onSubmit={handleSubmit} className="crazy-card-user-form">
      <label>
        <span>Title:</span>
        <Select
          value={titleOptions.find((option) => option.value === user.title)}
          name="title"
          options={titleOptions}
          onChange={({ value }) => {
            handleChange('title', value);
          }}
        />
      </label>

      <label>
        <span>Firstname:</span>
        <input
          className="crazy-card-user-form__input"
          name="firstname"
          type="text"
          value={user.firstname}
          onChange={({ target: { value } }) => handleChange('firstname', value)}
        />
      </label>

      <label>
        <span>Lastname:</span>
        <input
          className="crazy-card-user-form__input"
          name="lastname"
          type="text"
          value={user.lastname}
          onChange={({ target: { value } }) => handleChange('lastname', value)}
        />
      </label>

      <label>
        <span>Data of birth:</span>
        <DatePicker
          name="dateOfBirth"
          selected={user.dateOfBirth}
          dateFormat="dd/MM/yyyy"
          onChange={(date) => setDateOfBirth(date)}
          onChange={(date) => handleChange('dateOfBirth', date)}
          maxDate={getToday()}
          className="crazy-card-user-form__input"
          placeholderText="dd/MM/yyyy"
        />
      </label>

      <label>
        <span>Employment Status:</span>
        <Select
          value={employmentStatusOptions.find(
            (option) => option.value === user.employmentStatus
          )}
          name="employmentStatus"
          options={employmentStatusOptions}
          onChange={(option) => {
            handleChange('employmentStatus', option.value);
          }}
        />
      </label>

      <label>
        <span>Annual Income:</span>
        <input
          className="crazy-card-user-form__input"
          name="annualIncome"
          type="number"
          value={user.annualIncome}
          onChange={({ target: { value } }) =>
            handleChange('annualIncome', Number(value))
          }
        />
      </label>

      <label>
        <span>House Number:</span>
        <input
          className="crazy-card-user-form__input"
          name="houseNumber"
          type="number"
          value={user.houseNumber}
          onChange={({ target: { value } }) =>
            handleChange('houseNumber', Number(value))
          }
        />
      </label>

      <label>
        <span>Postcode:</span>
        <input
          className="crazy-card-user-form__input"
          name="postcode"
          type="text"
          value={user.postcode}
          onChange={({ target: { value } }) => handleChange('postcode', value)}
        />
      </label>

      <div className="crazy-card-user-form__actions">
        <button
          className="crazy-card-user-form__submit"
          type="button"
          onClick={(e) => handleSubmit(e)}
          disabled={checkForm(user)}
        >
          Submit
        </button>
      </div>

      {error && (
        <p className="crazy-card-user-form__error">
          All fields are required. Please fill in the form and submit again.
        </p>
      )}
    </form>
  );
};

UserForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
};

export default UserForm;
