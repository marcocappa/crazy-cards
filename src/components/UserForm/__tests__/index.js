import React from 'react';
import { shallow } from 'enzyme';
import UserForm from '../index';

// mock checkForm utility function
import { checkForm, getToday } from '../utility';
jest.mock('../utility');
getToday.mockReturnValue('2021-03-08');

describe('<UserForm />', () => {
  const onSubmit = jest.fn();
  const props = {
    onSubmit,
  };

  let wrapper;
  beforeEach(() => (wrapper = shallow(<UserForm {...props} />)));

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle inputs changes', () => {
    const title = 'Mr.';
    const firstname = 'FirstName';
    const lastname = 'LastName';
    const employmentStatus = 'Student';
    const annualIncome = 10000;
    const houseNumber = 100;
    const postcode = 'E2 37P';

    // title select
    wrapper.find('[name="title"]').simulate('change', { value: title });
    expect(wrapper.find('[name="title"]').props().value).toEqual({
      label: title,
      value: title,
    });

    // firstname
    wrapper
      .find('[name="firstname"]')
      .simulate('change', { target: { value: firstname } });
    expect(wrapper.find('[name="firstname"]').props().value).toEqual(firstname);

    // lastname
    wrapper
      .find('[name="lastname"]')
      .simulate('change', { target: { value: lastname } });
    expect(wrapper.find('[name="lastname"]').props().value).toEqual(lastname);

    // employmentStatus select
    wrapper
      .find('[name="employmentStatus"]')
      .simulate('change', { value: employmentStatus });
    expect(wrapper.find('[name="employmentStatus"]').props().value).toEqual({
      label: employmentStatus,
      value: employmentStatus,
    });

    //annualIncome
    wrapper
      .find('[name="annualIncome"]')
      .simulate('change', { target: { value: annualIncome } });
    expect(wrapper.find('[name="annualIncome"]').props().value).toEqual(
      annualIncome
    );

    // houseNumber
    wrapper
      .find('[name="houseNumber"]')
      .simulate('change', { target: { value: houseNumber } });
    expect(wrapper.find('[name="houseNumber"]').props().value).toEqual(
      houseNumber
    );

    // postcode
    wrapper
      .find('[name="postcode"]')
      .simulate('change', { target: { value: postcode } });
    expect(wrapper.find('[name="postcode"]').props().value).toEqual(postcode);
  });

  it('should invoke onSubmit', () => {
    checkForm.mockReturnValueOnce(true).mockReturnValueOnce(false);
    const preventDefault = jest.fn();

    // no error message on the form at the beginning
    expect(wrapper.exists('.crazy-card-user-form__error')).toEqual(false);

    // on submit:
    wrapper.find('button').simulate('click', { preventDefault });

    // invoke preventDefault, find error message avoid to call onSubmit
    expect(preventDefault).toHaveBeenCalledTimes(1);
    expect(onSubmit).toHaveBeenCalledTimes(0);
    expect(wrapper.exists('.crazy-card-user-form__error')).toEqual(true);

    // invoke preventDefault and onSubmit with no errors
    wrapper.find('button').simulate('click', { preventDefault });
    expect(preventDefault).toHaveBeenCalledTimes(2);
    expect(onSubmit).toHaveBeenCalledTimes(1);
    expect(wrapper.exists('.crazy-card-user-form__error')).toEqual(false);
  });
});
