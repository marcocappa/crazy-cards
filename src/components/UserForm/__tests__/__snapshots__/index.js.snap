// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserForm /> should match snapshot 1`] = `
<form
  className="crazy-card-user-form"
  onSubmit={[Function]}
>
  <label>
    <span>
      Title:
    </span>
    <StateManager
      defaultInputValue=""
      defaultMenuIsOpen={false}
      defaultValue={null}
      name="title"
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "Mr.",
            "value": "Mr.",
          },
          Object {
            "label": "Mrs.",
            "value": "Mrs.",
          },
          Object {
            "label": "Miss",
            "value": "Miss",
          },
        ]
      }
    />
  </label>
  <label>
    <span>
      Firstname:
    </span>
    <input
      className="crazy-card-user-form__input"
      name="firstname"
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <label>
    <span>
      Lastname:
    </span>
    <input
      className="crazy-card-user-form__input"
      name="lastname"
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <label>
    <span>
      Data of birth:
    </span>
    <r
      allowSameDay={false}
      className="crazy-card-user-form__input"
      customTimeInput={null}
      dateFormat="dd/MM/yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      enableTabLoop={true}
      excludeScrollbar={true}
      focusSelectedMonth={false}
      maxDate="2021-03-08"
      monthsShown={1}
      name="dateOfBirth"
      nextMonthButtonLabel="Next Month"
      nextYearButtonLabel="Next Year"
      onBlur={[Function]}
      onCalendarClose={[Function]}
      onCalendarOpen={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      placeholderText="dd/MM/yyyy"
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      previousYearButtonLabel="Previous Year"
      readOnly={false}
      renderDayContents={[Function]}
      selected=""
      shouldCloseOnSelect={true}
      showFourColumnMonthYearPicker={false}
      showFullMonthYearPicker={false}
      showMonthYearPicker={false}
      showPopperArrow={true}
      showPreviousMonths={false}
      showQuarterYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      showTwoColumnMonthYearPicker={false}
      showYearPicker={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      withPortal={false}
      yearItemNumber={12}
    />
  </label>
  <label>
    <span>
      Employment Status:
    </span>
    <StateManager
      defaultInputValue=""
      defaultMenuIsOpen={false}
      defaultValue={null}
      name="employmentStatus"
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "Student",
            "value": "Student",
          },
          Object {
            "label": "Part Time",
            "value": "Part Time",
          },
          Object {
            "label": "Full Time",
            "value": "Full Time",
          },
        ]
      }
    />
  </label>
  <label>
    <span>
      Annual Income:
    </span>
    <input
      className="crazy-card-user-form__input"
      name="annualIncome"
      onChange={[Function]}
      type="number"
      value={0}
    />
  </label>
  <label>
    <span>
      House Number:
    </span>
    <input
      className="crazy-card-user-form__input"
      name="houseNumber"
      onChange={[Function]}
      type="number"
      value={0}
    />
  </label>
  <label>
    <span>
      Postcode:
    </span>
    <input
      className="crazy-card-user-form__input"
      name="postcode"
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <div
    className="crazy-card-user-form__actions"
  >
    <button
      className="crazy-card-user-form__submit"
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </div>
</form>
`;
